// async-i2c-bus/dist/main/lib/Bus.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const bluebird_1 = require("bluebird");
const i2c_bus_1 = require("i2c-bus");
const BusError_1 = __importDefault(require("./BusError"));
const Bus = ({ busNumber = 1, openBus = i2c_bus_1.open } = {}) => {
    let i2cBus = null;
    let isOpen = false;
    return {
        get busNumber() {
            return busNumber;
        },
        get i2cBus() {
            return i2cBus;
        },
        get isOpen() {
            return isOpen;
        },
        open() {
            return new Promise((resolve, reject) => {
                i2cBus = bluebird_1.promisifyAll(openBus(this.busNumber, (error) => {
                    if (error) {
                        reject(`Error opening i2c bus: ${error.message}`);
                    }
                    isOpen = true;
                    resolve(this);
                }));
            });
        },
        async close() {
            if (!i2cBus || !isOpen) {
                throw new BusError_1.default('Bus is not open', busNumber);
            }
            await i2cBus.closeAsync();
            isOpen = false;
            i2cBus = null;
            return this;
        },
        async i2cFuncs() {
            if (!i2cBus || !isOpen) {
                throw new BusError_1.default('Bus is not open', busNumber);
            }
            return i2cBus.i2cFuncsAsync();
        },
        async scan() {
            if (!i2cBus || !isOpen) {
                throw new BusError_1.default('Bus is not open', busNumber);
            }
            return i2cBus.scanAsync();
        },
        async i2cRead(address, length, buffer) {
            if (!i2cBus || !isOpen) {
                throw new BusError_1.default('Bus is not open', busNumber);
            }
            return i2cBus.i2cReadAsync(address, length, buffer);
        },
        async i2cWrite(address, length, buffer) {
            if (!i2cBus || !isOpen) {
                throw new BusError_1.default('Bus is not open', busNumber);
            }
            return i2cBus.i2cWriteAsync(address, length, buffer);
        },
        async receiveByte(address) {
            if (!i2cBus || !isOpen) {
                throw new BusError_1.default('Bus is not open', busNumber);
            }
            return i2cBus.receiveByteAsync(address);
        },
        async sendByte(address, byte) {
            if (!i2cBus || !isOpen) {
                throw new BusError_1.default('Bus is not open', busNumber);
            }
            return i2cBus.sendByteAsync(address, byte);
        },
        async readByte(address, command) {
            if (!i2cBus || !isOpen) {
                throw new BusError_1.default('Bus is not open', busNumber);
            }
            return i2cBus.readByteAsync(address, command);
        },
        async readWord(address, command) {
            if (!i2cBus || !isOpen) {
                throw new BusError_1.default('Bus is not open', busNumber);
            }
            return i2cBus.readWordAsync(address, command);
        },
        async readI2cBlock(address, command, length, buffer) {
            if (!i2cBus || !isOpen) {
                throw new BusError_1.default('Bus is not open', busNumber);
            }
            return i2cBus.readI2cBlockAsync(address, command, length, buffer);
        },
        async writeByte(address, command, byte) {
            if (!i2cBus || !isOpen) {
                throw new BusError_1.default('Bus is not open', busNumber);
            }
            return i2cBus.writeByteAsync(address, command, byte);
        },
        async writeWord(address, command, word) {
            if (!i2cBus || !isOpen) {
                throw new BusError_1.default('Bus is not open', busNumber);
            }
            return i2cBus.writeWordAsync(address, command, word);
        },
        async writeI2cBlock(address, command, length, buffer) {
            if (!i2cBus || !isOpen) {
                throw new BusError_1.default('Bus is not open', busNumber);
            }
            return i2cBus.writeI2cBlockAsync(address, command, length, buffer);
        },
    };
};
exports.default = Bus;
