// async-i2c-bus/dist/main/lib/createI2cBusMock.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const defaultI2cBusFuncs = {
    i2c: false,
    tenBitAddr: false,
    protocolMangling: false,
    smbusPec: false,
    smbusBlockProcCall: false,
    smbusQuick: false,
    smbusReceiveByte: false,
    smbusSendByte: false,
    smbusReadByte: false,
    smbusWriteByte: false,
    smbusReadWord: false,
    smbusWriteWord: false,
    smbusProcCall: false,
    smbusReadBlock: false,
    smbusWriteBlock: false,
    smbusReadI2cBlock: false,
    smbusWriteI2cBlock: false,
};
// @TODO Add full features
const createI2cBusMock = ({ devices = {}, i2cBusFuncs = defaultI2cBusFuncs, } = {}) => (_, openCallback) => {
    openCallback(null);
    return {
        close(callback) {
            callback(null);
        },
        closeSync() {
            return;
        },
        i2cFuncsSync() {
            return i2cBusFuncs;
        },
        i2cFuncs(callback) {
            callback(null, this.i2cFuncsSync());
        },
        scanSync() {
            return Object.keys(devices).map(address => parseInt(address, 10));
        },
        scan(callback) {
            callback(null, this.scanSync());
        },
        i2cReadSync(address, length, buffer) {
            return devices[address].copy(buffer, 0, 0, length);
        },
        i2cRead(address, length, buffer, callback) {
            callback(null, this.i2cReadSync(address, length, buffer), buffer);
        },
        i2cWriteSync(address, length, buffer) {
            return buffer.copy(devices[address], 0, 0, length);
        },
        i2cWrite(address, length, buffer, callback) {
            callback(null, this.i2cWriteSync(address, length, buffer), buffer);
        },
        readByteSync(address, command) {
            return devices[address][command];
        },
        readByte(address, command, callback) {
            callback(null, this.readByteSync(address, command));
        },
        readWordSync(address, command) {
            return devices[address].readUInt16BE(command);
        },
        readWord(address, command, callback) {
            return callback(null, this.readWordSync(address, command));
        },
        readI2cBlockSync(address, command, length, buffer) {
            return devices[address].copy(buffer, 0, command, command + length);
        },
        readI2cBlock(address, command, length, buffer, callback) {
            callback(null, this.readI2cBlockSync(address, command, length, buffer), buffer);
        },
        receiveByteSync(address) {
            return this.readByteSync(address, 0);
        },
        receiveByte(address, callback) {
            callback(null, this.receiveByteSync(address));
        },
        writeByteSync(address, command, byte) {
            devices[address][command] = byte;
        },
        writeByte(address, command, byte, callback) {
            this.writeByteSync(address, command, byte);
            callback(null);
        },
        sendByteSync(address, byte) {
            this.writeByteSync(address, 0, byte);
        },
        sendByte(address, byte, callback) {
            this.sendByteSync(address, byte);
            callback(null);
        },
        writeWordSync(address, command, word) {
            devices[address].writeUInt16BE(word, command);
        },
        writeWord(address, command, word, callback) {
            this.writeByteSync(address, command, word);
            callback(null);
        },
        writeQuickSync() {
            return null;
        },
        writeQuick(__, ___, ____, callback) {
            callback(null);
        },
        writeI2cBlockSync(address, command, length, buffer) {
            return buffer.copy(devices[address], command, 0, length);
        },
        writeI2cBlock(address, command, length, buffer, callback) {
            callback(null, this.writeI2cBlockSync(address, command, length, buffer), buffer);
        },
    };
};
exports.default = createI2cBusMock;
