// async-i2c-bus/dist/main/lib/Device.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Device = ({ address, bus }) => ({
    get address() {
        return address;
    },
    get bus() {
        return bus;
    },
    i2cRead(length, buffer) {
        return bus.i2cRead(address, length, buffer);
    },
    i2cWrite(length, buffer) {
        return bus.i2cWrite(address, length, buffer);
    },
    receiveByte() {
        return bus.receiveByte(address);
    },
    sendByte(byte) {
        return bus.sendByte(address, byte);
    },
    readByte(command) {
        return bus.readByte(address, command);
    },
    readWord(command) {
        return bus.readWord(address, command);
    },
    readI2cBlock(command, length, buffer) {
        return bus.readI2cBlock(address, command, length, buffer);
    },
    writeByte(command, byte) {
        return bus.writeByte(address, command, byte);
    },
    writeWord(command, word) {
        return bus.writeWord(address, command, word);
    },
    writeI2cBlock(command, length, buffer) {
        return bus.writeI2cBlock(address, command, length, buffer);
    },
});
exports.default = Device;
