// async-i2c-bus/dist/module/lib/Device.js
export default class Device {
    constructor({ bus, address }) {
        this.bus = bus;
        this.address = address;
    }
    async readByte(command) {
        const result = await this.bus.readByte(this.address, command);
        return result;
    }
    async readWord(command) {
        const result = await this.bus.readWord(this.address, command);
        return result;
    }
    async readI2cBlock(command, length, buffer) {
        const result = await this.bus.readI2cBlock(this.address, command, length, buffer);
        return result;
    }
    async receiveByte() {
        const result = await this.bus.receiveByte(this.address);
        return result;
    }
    async sendByte(byte) {
        await this.bus.sendByte(this.address, byte);
        return this;
    }
    async writeByte(command, byte) {
        await this.bus.writeByte(this.address, command, byte);
        return this;
    }
    async writeWord(command, word) {
        await this.bus.writeWord(this.address, command, word);
        return this;
    }
    async writeQuick(bit) {
        await this.bus.writeQuick(this.address, bit);
        return this;
    }
    async writeI2cBlock(command, length, buffer) {
        const result = await this.bus.writeI2cBlock(this.address, command, length, buffer);
        return result;
    }
    async i2cRead(length, buffer) {
        const result = await this.bus.i2cRead(this.address, length, buffer);
        return result;
    }
    async i2cWrite(length, buffer) {
        const result = await this.bus.i2cWrite(this.address, length, buffer);
        return result;
    }
    async deviceId() {
        const result = await this.bus.deviceId(this.address);
        return result;
    }
    static create({ bus, address }) {
        return new Device({ bus, address });
    }
}
