// async-i2c-bus/dist/main/lib/Bus.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const i2c_bus_1 = require("i2c-bus");
const BusError_1 = __importDefault(require("./BusError"));
class Bus {
    constructor({ busNumber }) {
        this.isOpen = false;
        this.busNumber = busNumber;
    }
    guardAgainstClosedBus() {
        if (!this.isOpen) {
            throw BusError_1.default.create('Bus is not open');
        }
    }
    async open() {
        try {
            if (!this.isOpen) {
                this.i2cBus = await i2c_bus_1.openPromisified(this.busNumber);
                this.isOpen = true;
            }
            return this;
        }
        catch (error) {
            throw BusError_1.default.createFromError(error);
        }
    }
    async close() {
        this.guardAgainstClosedBus();
        try {
            await this.i2cBus.close();
            this.isOpen = false;
            return this;
        }
        catch (error) {
            throw BusError_1.default.createFromError(error);
        }
    }
    async i2cFuncs() {
        this.guardAgainstClosedBus();
        try {
            const result = await this.i2cBus.i2cFuncs();
            return result;
        }
        catch (error) {
            throw BusError_1.default.createFromError(error);
        }
    }
    async readByte(address, command) {
        this.guardAgainstClosedBus();
        try {
            const result = await this.i2cBus.readByte(address, command);
            return result;
        }
        catch (error) {
            throw BusError_1.default.createFromError(error);
        }
    }
    async readWord(address, command) {
        this.guardAgainstClosedBus();
        try {
            const result = await this.i2cBus.readWord(address, command);
            return result;
        }
        catch (error) {
            throw BusError_1.default.createFromError(error);
        }
    }
    async readI2cBlock(address, command, length, buffer) {
        this.guardAgainstClosedBus();
        try {
            const result = await this.i2cBus.readI2cBlock(address, command, length, buffer);
            return result;
        }
        catch (error) {
            throw BusError_1.default.createFromError(error);
        }
    }
    async receiveByte(address) {
        this.guardAgainstClosedBus();
        try {
            const result = await this.i2cBus.receiveByte(address);
            return result;
        }
        catch (error) {
            throw BusError_1.default.createFromError(error);
        }
    }
    async sendByte(address, byte) {
        this.guardAgainstClosedBus();
        try {
            await this.i2cBus.sendByte(address, byte);
            return this;
        }
        catch (error) {
            throw BusError_1.default.createFromError(error);
        }
    }
    async writeByte(address, command, byte) {
        this.guardAgainstClosedBus();
        try {
            await this.i2cBus.writeByte(address, command, byte);
            return this;
        }
        catch (error) {
            throw BusError_1.default.createFromError(error);
        }
    }
    async writeWord(address, command, word) {
        this.guardAgainstClosedBus();
        try {
            await this.i2cBus.writeWord(address, command, word);
            return this;
        }
        catch (error) {
            throw BusError_1.default.createFromError(error);
        }
    }
    async writeQuick(address, bit) {
        this.guardAgainstClosedBus();
        try {
            await this.i2cBus.writeQuick(address, bit);
            return this;
        }
        catch (error) {
            throw BusError_1.default.createFromError(error);
        }
    }
    async writeI2cBlock(address, command, length, buffer) {
        this.guardAgainstClosedBus();
        try {
            const result = await this.i2cBus.writeI2cBlock(address, command, length, buffer);
            return result;
        }
        catch (error) {
            throw BusError_1.default.createFromError(error);
        }
    }
    async i2cRead(address, length, buffer) {
        this.guardAgainstClosedBus();
        try {
            const result = await this.i2cBus.i2cRead(address, length, buffer);
            return result;
        }
        catch (error) {
            throw BusError_1.default.createFromError(error);
        }
    }
    async i2cWrite(address, length, buffer) {
        this.guardAgainstClosedBus();
        try {
            const result = await this.i2cBus.i2cWrite(address, length, buffer);
            return result;
        }
        catch (error) {
            throw BusError_1.default.createFromError(error);
        }
    }
    async scan(startAddress, endAddress) {
        this.guardAgainstClosedBus();
        try {
            const result = await this.i2cBus.scan(startAddress, endAddress);
            return result;
        }
        catch (error) {
            throw BusError_1.default.createFromError(error);
        }
    }
    async deviceId(address) {
        this.guardAgainstClosedBus();
        try {
            const result = await this.i2cBus.deviceId(address);
            return result;
        }
        catch (error) {
            throw BusError_1.default.createFromError(error);
        }
    }
    static create({ busNumber }) {
        return new Bus({ busNumber });
    }
    static createAndOpen({ busNumber }) {
        const bus = new Bus({ busNumber });
        return bus.open();
    }
}
exports.default = Bus;
