// async-bmp280/dist/main/lib/BMP280.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const async_i2c_bus_1 = require("async-i2c-bus");
const constants_1 = require("./constants");
const helpers_1 = require("./helpers");
const utils_1 = require("./utils");
function BMP280({ address = constants_1.ADDRESS, bus }) {
    const temperatureCorrection = [0, 0, 0];
    const pressureCorrection = [0, 0, 0, 0, 0, 0, 0, 0, 0];
    return Object.assign({}, async_i2c_bus_1.Device({ address, bus }), { async init({ temperatureOversampling = 'x1', pressureOversampling = 'x4', mode = 'NORMAL', standbyTime = '500us', iirFilter = 'x16', } = {}) {
            const id = await this.getId();
            if (id !== constants_1.ID) {
                throw new async_i2c_bus_1.DeviceError('Id does not correspond with a BMP280', this.bus.busNumber, constants_1.ADDRESS);
            }
            await this.reset();
            await Promise.all([this.readTemperatureCorrection(), this.readPressureCorrection()]);
            await this.writeControlMeasurement({
                temperatureOversampling,
                pressureOversampling,
                mode,
            });
            await this.writeConfig({
                standbyTime,
                iirFilter,
            });
            return this;
        },
        async getId() {
            return this.readByte(constants_1.REGISTERS.ID);
        },
        async reset() {
            await this.writeByte(constants_1.REGISTERS.RESET, 0xb6);
            return this;
        },
        async readStatus() {
            const status = await this.readByte(constants_1.REGISTERS.STATUS);
            return {
                measuring: !!utils_1.extractValue(status, constants_1.MASKS.MEASURING, constants_1.OFFSETS.MEASURING),
                imUpdate: !!utils_1.extractValue(status, constants_1.MASKS.IM_UPDATE, constants_1.OFFSETS.IM_UPDATE),
            };
        },
        async readControlMeasurement() {
            const ctrlMeas = await this.readByte(constants_1.REGISTERS.CTRL_MEAS);
            const { osrsT, osrsP, mode } = helpers_1.extractCtrlMeas(ctrlMeas);
            return {
                temperatureOversampling: helpers_1.readOversampling(osrsT),
                pressureOversampling: helpers_1.readOversampling(osrsP),
                mode: helpers_1.readMode(mode),
            };
        },
        async writeControlMeasurement(controlMeasurement) {
            const currentValue = await this.readByte(constants_1.REGISTERS.CTRL_MEAS);
            const { osrsT: currentOsrsT, osrsP: currentOsrsP, mode: currentMode } = helpers_1.extractCtrlMeas(currentValue);
            const osrsT = helpers_1.parseOversampling(controlMeasurement.temperatureOversampling, currentOsrsT);
            const osrsP = helpers_1.parseOversampling(controlMeasurement.pressureOversampling, currentOsrsP);
            const mode = helpers_1.parseMode(controlMeasurement.mode, currentMode);
            await this.writeByte(constants_1.REGISTERS.CTRL_MEAS, (osrsT << constants_1.OFFSETS.OSRS_T) | (osrsP << constants_1.OFFSETS.OSRS_P) | (mode << constants_1.OFFSETS.MODE));
            return this;
        },
        async readConfig() {
            const config = await this.readByte(constants_1.REGISTERS.CONFIG);
            const { tSb, filter } = helpers_1.extractConfig(config);
            return {
                standbyTime: helpers_1.readStandbyTime(tSb),
                iirFilter: helpers_1.readIirFilter(filter),
            };
        },
        async writeConfig(config) {
            const currentValue = await this.readByte(constants_1.REGISTERS.CONFIG);
            const { tSb: currentTsB, filter: currentFilter } = helpers_1.extractConfig(currentValue);
            const tSb = helpers_1.parseStandbyTime(config.standbyTime, currentTsB);
            const filter = helpers_1.parseIirFilter(config.iirFilter, currentFilter);
            await this.writeByte(constants_1.REGISTERS.CONFIG, (tSb << constants_1.OFFSETS.T_SB) | (filter << constants_1.OFFSETS.FILTER));
            return this;
        },
        async readTemperatureCorrection() {
            const buffer = Buffer.alloc(6);
            await this.readI2cBlock(constants_1.REGISTERS.TEMP_CORRECTION, 6, buffer);
            temperatureCorrection[0] = buffer.readUInt16LE(0);
            temperatureCorrection[1] = buffer.readInt16LE(2);
            temperatureCorrection[2] = buffer.readInt16LE(4);
            return this;
        },
        async readPressureCorrection() {
            const buffer = Buffer.alloc(18);
            await this.readI2cBlock(constants_1.REGISTERS.PRESS_CORRECTION, 18, buffer);
            pressureCorrection[0] = buffer.readUInt16LE(0);
            pressureCorrection[1] = buffer.readInt16LE(2);
            pressureCorrection[2] = buffer.readInt16LE(4);
            pressureCorrection[3] = buffer.readInt16LE(6);
            pressureCorrection[4] = buffer.readInt16LE(8);
            pressureCorrection[5] = buffer.readInt16LE(10);
            pressureCorrection[6] = buffer.readInt16LE(12);
            pressureCorrection[7] = buffer.readInt16LE(14);
            pressureCorrection[8] = buffer.readInt16LE(16);
            return this;
        },
        async readTemperature() {
            const buffer = Buffer.alloc(3);
            await this.readI2cBlock(constants_1.REGISTERS.TEMP, 3, buffer);
            const temperature = helpers_1.calculateTemperature(buffer.readUIntBE(0, 3) >>> 4, temperatureCorrection);
            return temperature / 5120;
        },
        async readPressure() {
            const tempBuffer = Buffer.alloc(3);
            const pressBuffer = Buffer.alloc(3);
            await Promise.all([
                this.readI2cBlock(constants_1.REGISTERS.TEMP, 3, tempBuffer),
                this.readI2cBlock(constants_1.REGISTERS.PRESS, 3, pressBuffer),
            ]);
            const temperature = helpers_1.calculateTemperature(tempBuffer.readUIntBE(0, 3) >>> 4, temperatureCorrection);
            return helpers_1.calculatePressure(pressBuffer.readUIntBE(0, 3) >>> 4, temperature, pressureCorrection);
        } });
}
exports.default = BMP280;
