// async-bmp280/dist/main/lib/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const constants_1 = require("./constants");
const utils_1 = require("./utils");
/**
 * @hidden
 */
exports.extractCtrlMeas = (value) => ({
    osrsT: utils_1.extractValue(value, constants_1.MASKS.OSRS_T, constants_1.OFFSETS.OSRS_T),
    osrsP: utils_1.extractValue(value, constants_1.MASKS.OSRS_P, constants_1.OFFSETS.OSRS_P),
    mode: utils_1.extractValue(value, constants_1.MASKS.MODE, constants_1.OFFSETS.MODE),
});
/**
 * @hidden
 */
exports.extractConfig = (value) => ({
    tSb: utils_1.extractValue(value, constants_1.MASKS.T_SB, constants_1.OFFSETS.T_SB),
    filter: utils_1.extractValue(value, constants_1.MASKS.FILTER, constants_1.OFFSETS.FILTER),
});
/**
 * @hidden
 */
exports.parseOversampling = utils_1.lookUp(constants_1.OVERSAMPLING);
/**
 * @hidden
 */
exports.readOversampling = utils_1.reverseLookUp(constants_1.OVERSAMPLING);
/**
 * @hidden
 */
exports.parseMode = utils_1.lookUp(constants_1.MODE);
/**
 * @hidden
 */
exports.readMode = utils_1.reverseLookUp(constants_1.MODE);
/**
 * @hidden
 */
exports.parseStandbyTime = utils_1.lookUp(constants_1.STANDBY_TIME);
/**
 * @hidden
 */
exports.readStandbyTime = utils_1.reverseLookUp(constants_1.STANDBY_TIME);
/**
 * @hidden
 */
exports.parseIirFilter = utils_1.lookUp(constants_1.IIR_FILTER);
/**
 * @hidden
 */
exports.readIirFilter = utils_1.reverseLookUp(constants_1.IIR_FILTER);
/**
 * @hidden
 */
exports.calculateTemperature = (rawTemperature, correction) => {
    const var1 = (rawTemperature / 16384 - correction[0] / 1024) * correction[1];
    const var2 = Math.pow(rawTemperature / 131072 - correction[0] / 8192, 2) * correction[2];
    return var1 + var2;
};
/**
 * @hidden
 */
exports.calculatePressure = (rawPressure, temperatureFine, correction) => {
    let var1 = temperatureFine / 2 - 64000;
    let var2 = var1 * (var1 * (correction[5] / 32768) + correction[4] * 2);
    var2 = var2 / 4 + correction[3] * 65536;
    var1 = ((correction[2] * var1 * var1) / 524288 + correction[1] * var1) / 524288;
    var1 = (1 + var1 / 32768) * correction[0];
    if (var1 === 0) {
        return 0;
    }
    let var3 = 1048576 - rawPressure;
    var3 = (var3 - var2 / 4096) * (6250 / var1);
    var1 = correction[8] * ((var3 * var3) / 2147483648);
    var2 = (var3 * correction[7]) / 32768;
    return var3 + (var1 + var2 + correction[6]) / 16;
};
